File: src/android/app_lovin/src/main/java/com/ee/internal/AppLovinInterstitialAdListener.kt
package com.ee.internal

import androidx.annotation.AnyThread
import com.applovin.sdk.AppLovinAd
import com.applovin.sdk.AppLovinAdClickListener
import com.applovin.sdk.AppLovinAdDisplayListener
import com.applovin.sdk.AppLovinAdLoadListener
import com.ee.ILogger
import com.ee.IMessageBridge
import com.ee.Thread
import java.util.concurrent.atomic.AtomicBoolean

internal class AppLovinInterstitialAdListener(
    private val _bridge: IMessageBridge,
    private val _logger: ILogger)
    : AppLovinAdLoadListener, AppLovinAdDisplayListener, AppLovinAdClickListener {
    companion object {
        private val kTag = AppLovinInterstitialAdListener::class.java.name
        private const val kPrefix = "AppLovinBridge"
        private const val kOnInterstitialAdLoaded = "${kPrefix}OnInterstitialAdLoaded"
        private const val kOnInterstitialAdFailedToLoad = "${kPrefix}OnInterstitialAdFailedToLoad"
        private const val kOnInterstitialAdClicked = "${kPrefix}OnInterstitialAdClicked"
        private const val kOnInterstitialAdClosed = "${kPrefix}OnInterstitialAdClosed"
    }

    private val _isLoaded = AtomicBoolean(false)

    val isLoaded: Boolean
        @AnyThread get() = _isLoaded.get()

    override fun adReceived(ad: AppLovinAd) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::adReceived.name}")
            _isLoaded.set(true)
            _bridge.callCpp(kOnInterstitialAdLoaded)
        }
    }

    override fun failedToReceiveAd(errorCode: Int) {
        Thread.runOnMainThread {
            _logger.info("$kTag: ${this::failedToReceiveAd.name}: code $errorCode")
            _bridge.callCpp(kOnInterstitialAdFailedToLoad, errorCode.toString())
        }
    }

    override fun adDisplayed(ad: AppLovinAd) {
        Thread.runOnMainThread {
            _logger.info("$kTag: ${this::adDisplayed.name}")
        }
    }

    override fun adClicked(ad: AppLovinAd) {
        Thread.runOnMainThread {
            _logger.info("$kTag: ${this::adClicked.name}")
            _bridge.callCpp(kOnInterstitialAdClicked)
        }
    }

    override fun adHidden(ad: AppLovinAd) {
        Thread.runOnMainThread {
            _logger.info(this::adHidden.name)
            _isLoaded.set(false)
            _bridge.callCpp(kOnInterstitialAdClosed)
        }
    }
}

File: src/ios/ee/app_lovin/AppLovinInterstitialAdListener.swift
//
//  AppLovinInterstitialAdListener.swift
//  Pods
//
//  Created by eps on 6/26/20.
//

import AppLovinSDK

private let kTag = "\(AppLovinInterstitialAdListener.self)"
private let kPrefix = "AppLovinBridge"
private let kOnInterstitialAdLoaded = "\(kPrefix)OnInterstitialAdLoaded"
private let kOnInterstitialAdFailedToLoad = "\(kPrefix)OnInterstitialAdFailedToLoad"
private let kOnInterstitialAdClicked = "\(kPrefix)OnInterstitialAdClicked"
private let kOnInterstitialAdClosed = "\(kPrefix)OnInterstitialAdClosed"

internal class AppLovinInterstitialAdListener:
    NSObject, ALAdLoadDelegate, ALAdDisplayDelegate {
    private let _bridge: IMessageBridge
    private let _logger: ILogger
    private var _isLoaded = false

    init(_ bridge: IMessageBridge, _ logger: ILogger) {
        _bridge = bridge
        _logger = logger
        super.init()
    }

    var isLoaded: Bool {
        return _isLoaded
    }

    func adService(_ adService: ALAdService, didLoad ad: ALAd) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function)")
            self._isLoaded = true
            self._bridge.callCpp(kOnInterstitialAdLoaded)
        }
    }

    func adService(_ adService: ALAdService, didFailToLoadAdWithError code: Int32) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function): code \(code)")
            self._bridge.callCpp(kOnInterstitialAdFailedToLoad, EEJsonUtils.convertDictionary(toString: [
                "code": code,
                "message": ""
            ]))
        }
    }

    func ad(_ ad: ALAd, wasDisplayedIn view: UIView) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function)")
            self._isLoaded = false
        }
    }

    func ad(_ ad: ALAd, wasClickedIn view: UIView) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function)")
            self._bridge.callCpp(kOnInterstitialAdClicked)
        }
    }

    func ad(_ ad: ALAd, wasHiddenIn view: UIView) {
        Thread.runOnMainThread {
            self._logger.debug("\(kTag): \(#function)")
            self._bridge.callCpp(kOnInterstitialAdClosed)
        }
    }
}
